### ISSUE TITLE
Generic type is not correctly detected

### ISSUE THREAD

**Issue by tomitrescak** [microsoft/TypeScript#41809]
**TypeScript Version:**  4.1.2

**Search Terms:**

**Expected behavior:**

The generic type for props V1 & V2 in the code below is correctly detected

**Actual behavior:**

The generic type for props V1 & V2 in the code below is not detected and marked as undefined.

**Related Issues:**

**Code**
```ts
type Component<A> = {
    (): void;
}

type QueryResult<A, B> = {}

type Options<B> = {
    variables: B
}

type Q1 = { id1: string };
type V1 = { id1: string };
type Q2 = { id2: string };
type V2 = { id2: string };

// this is the problematic definition, the props there are not correctly detected
function withQueries<Q1, V1, Q2, V2>(queryProps: {
  queries: (props: V1 & V2) => [
      QueryResult<Q1, V1>, 
      QueryResult<Q2, V2>
  ];
}): Component<V1 & V2> {
    return null as any;
}

let useQuery1: (options: Options<V1>) => QueryResult<Q1, V1>
let useQuery2: (options: Options<V2>) => QueryResult<Q2, V2>

const ScheduleContainer = withQueries({
  queries(p)  {
    return [
      useQuery1({ variables: { id1: p.id } }),   // <- p is unknown
      useQuery2({ variables: { id2: p.id } })   // <- p is unknown
    ];
  },
});
```

<details><summary><b>Output</b></summary>

```ts
"use strict";
function withQueries(queryProps) {
    return null;
}
let useQuery1;
let useQuery2;
const ScheduleContainer = withQueries({
    queries(p) {
        return [
            useQuery1({ variables: { id1: p.id } }),
            useQuery2({ variables: { id2: p.id } }) 
        ];
    },
});

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "alwaysStrict": true,
    "esModuleInterop": true,
    "declaration": true,
    "experimentalDecorators": true,
    "emitDecoratorMetadata": true,
    "moduleResolution": 2,
    "target": "ES2017",
    "jsx": "React",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play?ssl=18&ssc=2&pln=16&pc=1#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCoASMFmgY7p7ehCTkVOle4jUMdRLYTqCc8LEZzCZmoNJODPHFZZXVtZYYkAAeRJTInDQS8po+rf40gnjbDKv8LqD2jpbYoACqAEoAMsK7sUmxkGSCc+VVQQuaTwVb1UBrDYULY7PagbgUZLJH6QbYmJAECjuMigZEMVDsJzCFLNXxtajBBCcQQ0MwAUVWDEQNUgADVHBQGNJ3KAALygABEAAkYNAMOB4GRogLFFTBPB3AExcwABT0xnM9zsyhc9wASmCKhwDQ8ZC8iElzhB7Bo3zcZmY7AYzEg-Fg0HUiS58D0Ii8AoZTJZggFSRxAvADlQAHJhAA5SASAVBFQAeW+ZF2gldWkgx1QjgUrmkeFATgtOlGWH0KAQiBhwiudokkuiIgMHBx3RYbC43CCilQLVAAGEMfBOKjUAAeCoAPj5oAA3ooSiUVXr8qAdPAKNwANyKAC+wSHRFAAEV2D88E99FEZxV+AAhBf8pcnwfD1PUCgTwTTq+i4rmu24crq+hbs+x6nsOF4AIzASI3DwVuNKUJwYxHoeZ4oKyiHvshqHXKgGFYThcEAExIXulFoaRVbYV+56stRhG0fRZGgExijJFwsCoH+XgfJgV4-BQ+jTgh-D4fwF6UTJlFziqACO15kHgtBsAQBTLquoBqeJkGgCqBDabp+GgAAZKArF6nyC4ANr6aBYkaXeggPlJ8EyfBc78C5a5ube97QDO8mKXO+kALqHkem6juOk7bNOlk2axC4gWuTgEmagxRNAYiNpweBxcE7g0Ow2bBcRKrwL+-5bj+gn-qlfn2byC7BR5XnSbZfmKBVoBVZAwV0SZ9UtdSTUNdSqVKR1XXqSFnlhVJCm2UplL-jQpDYNwUSQGOZwZD8i4iW5EmCCqWWGZQ+imfZemgTlhJeM5oElCNNU3WBOrcsZS5EVuBABHu3HcXqvCBV91XLZRv06OBAO6UDHHVGDrhHpDgWxfpR7Q-F+5AA)

**Comment by tomitrescak** [comment]
I guess I know what is happening, since type of `p` isunknown it does not continue detecting the type from Queryresult
